plutus-ir/src/Language/PlutusIR/Analysis/Free.hs

```diff
@@ -0,0 +1,121 @@
+shown by the example:
+
+(let nonrec { x = y
+              y = 3 }
+) ===> y is free in x
+
+VS
+
+(let rec { x = y
+           y = 3 }
+) ===> y is not free in x
+
+-}
+
+
+-- | Given a term, it returns a set of all the FREE variables inside that term (i.e. not declared/scoped inside the term).
+--
+-- It does not require the input term to be prior 'PLC.rename'd.
+-- TODO: refactor using recursion-schemes (or usingtermSubterms, termSubtypes can be used and lens + fold), see 'language-plutus-core/src/Language/PlutusCore/Subst.hs'
+-- See Note: [PIR Free variables]
+fTerm :: forall name tyname a.
+  (Ord (tyname a), PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+  => Term tyname name a -> S.Set PLC.Unique
+fTerm = \case
+  Var _ x -> S.singleton $ x ^. PLC.unique . coerced
+
+  Let _ NonRec bs tIn -> fLet bs tIn NonRec
+  Let _ Rec bs tIn -> fLet bs tIn Rec
+
+  TyAbs _ n _ t -> S.delete (n ^. PLC.unique . coerced) $ fTerm t
+  LamAbs _ n ty t -> fType ty
+                     <>
+                     S.delete (n ^. PLC.unique . coerced) (fTerm t)
+  Apply _ t1 t2 -> fTerm t1 <> fTerm t2
+  TyInst _ t ty -> fTerm t <> fType ty
+  IWrap _ ty1 ty2 t -> mconcat [fType ty1, fType ty2, fTerm t]
+  Unwrap _ t -> fTerm t
+  _ -> S.empty
+
+ where
+  fLet :: [Binding tyname name a] -> Term tyname name a -> Recursivity -> S.Set PLC.Unique
+  fLet bs tIn =
+    let fIn = fTerm tIn S.\\ foldMap bindingIds bs -- the free variables of termIn (scoped by all the let id's)
+    in \case
+      Rec -> (foldMap fBinding bs S.\\ foldMap bindingIds bs) <> fIn -- all id's in the letrec can be seen (scoped) by each other
+      NonRec -> let initialScope = S.empty -- starting with an empty linear-scope
+                    initialFreeSet = fIn -- starting from the freeset found in tIn
+                in snd $ foldl nonRecVarAcc (initialScope, initialFreeSet) bs
+
+  -- it threads an ever-increasing linear-scope and an accumulated set of free variables found, i.e. (linearScope, freeSet)
+  -- See Note [Free variables and letnonrec linear scope]
+  -- See Note [Right-associative compilation of let-bindings for linear scoping]
+  nonRecVarAcc :: (S.Set PLC.Unique, S.Set PLC.Unique) -> Binding tyname name a -> (S.Set PLC.Unique, S.Set PLC.Unique)
+  nonRecVarAcc (accLinearScope,accFreeSet) b =
+    let newScope = bindingIds b <> accLinearScope
+        newFreeSet = (fBinding b S.\\ accLinearScope) <> accFreeSet
+    in (newScope, newFreeSet)
+
+-- | Given a PIR type, it returns the free type variables of that type.
+-- See Note: [PIR Free variables]
+fType :: (Ord (tyname a), PLC.HasUnique (tyname a) PLC.TypeUnique)
+      => Type tyname a -> S.Set PLC.Unique
+fType = S.map (^. PLC.unique . coerced) . ftvTy
+
+fVarDecl :: (Ord (tyname a), PLC.HasUnique (tyname a) PLC.TypeUnique)
+         => VarDecl tyname name a -> S.Set PLC.Unique
+fVarDecl (VarDecl _ _ ty) = fType ty
```

Comment:
Definitely worth adding a comment, because a function with that name might do something else (return the unique of the bound variable, return all the uniques, including the bound variable).